docs/topics/addons.rst

```diff
@@ -0,0 +1,145 @@
+    :class:`~scrapy.settings.Settings` object as wished, e.g. enable components
+    for this add-on or set required configuration of other extensions.
+
+    :param config: Configuration of this add-on
+    :type config: ``dict``
+
+    :param settings: The settings object storing Scrapy/component configuration
+    :type settings: :class:`~scrapy.settings.Settings`
+
+.. method:: check_configuration(config, crawler)
+
+    This method is called when the :class:`~scrapy.crawler.Crawler` has been
+    fully initialized, immediately before it starts crawling. You can perform
+    additional dependency and configuration checks here.
+
+    :param config: Configuration of this add-on
+    :type config: ``dict``
+
+    :param crawler: Fully initialized Scrapy crawler
+    :type crawler: :class:`~scrapy.crawler.Crawler`
+
+
+Add-on base class
+=================
+
+Scrapy comes with a built-in base class for add-ons which provides some
+convenience functionality: the add-on configuration can be exposed into
+Scrapy's settings via :meth:`~scrapy.addons.Addon.export_config`, configurable
+via :attr:`~scrapy.addons.Addon.default_config` and
+:attr:`~scrapy.addons.Addon.config_mapping`.
+
+By default, the base add-on class will expose the add-on configuration into
+Scrapy's settings namespace, in upper case. It is
+easy to write your own functionality while still being able to use the
+convenience functions by overwriting
+:meth:`~scrapy.addons.Addon.update_settings`.
+
+.. module:: scrapy.addons
+   :noindex:
+
+.. autoclass:: Addon
+   :members:
+
+
+Add-on examples
+===============
+
+Set some basic configuration using the :class:`Addon` base class::
+
+    from scrapy.addons import Addon
+
+    class MyAddon(Addon):
+        name = 'myaddon'
+
+        def update_settings(self, config, settings):
+            super().update_settings(settings)
+            settings["ITEM_PIPELINES"]["path.to.mypipeline"] = 200
+            settings["DNSCACHE_ENABLED"] = True
+
+Check dependencies::
+
+    from scrapy.addons import Addon
+
+    class MyAddon(Addon):
+        name = 'myaddon'
+
+        def update_settings(self, config, settings):
+            try:
+                import boto
+            except ImportError:
+                raise RuntimeError("myaddon requires the boto library")
+            super().update_settings(settings)
+
+Check configuration of fully initialized crawler (see
+:ref:`topics-api-crawler`)::
+
+    class MyAddon(object):
+        name = 'myaddon'
+
+        def update_settings(self, config, settings):
+            super().update_settings(settings)
+            settings.set('DNSCACHE_ENABLED', False, priority='addon')
+
+        def check_configuration(self, config, crawler):
+            if crawler.settings.getbool('DNSCACHE_ENABLED'):
+                # The spider, some other add-on, or the user messed with the
+                # DNS cache setting
+                raise ValueError("myaddon is incompatible with DNS cache")
```

Comment:
I am not sure whether or not we should encourage this approach of post-setting-update checking on add-ons. But this is one example where I think we don’t need a method. For example (untested):

```python
from logging import getLogger

from scrapy.signals import spider_opened

logger = getLogger(__name__)

class MinimalisticApproachAddon:
    
    @classmethod
    def from_crawler(cls, crawler):
        return cls(crawler)        

    def __init__(self, crawler):
        crawler.signals.connect(self.spider_opened, signal=spider_opened)

    def update_settings(self, settings):
        super().update_settings(settings)
        settings.set('DNSCACHE_ENABLED', False, priority='addon')
        
    def spider_opened(self, spider):
        if spider.settings.getbool('DNSCACHE_ENABLED'):
            logger.error(f"DNSCACHE_ENABLED is enabled, and {self.__class__} does not work with it.")
            return spider.crawler.engine.close_spider()
```